packages/amplify_api_rest/android/src/main/kotlin/com/amazonaws/amplify/amplify_api_rest/AmplifyApiRestPlugin.kt

```diff
@@ -0,0 +1,311 @@
+          )
+          operationsMap[code] = operation!!
+        }
+
+      } catch(e: Exception) {
+        prepareError(flutterResult, e, "hi")
+      }
+    }
+    else{
+      prepareError(flutterResult, java.lang.Exception(FlutterAuthFailureMessage.MALFORMED.toString()), FlutterAuthFailureMessage.MALFORMED.toString())
+    }
+
+  }
+
+  private fun prepareRestResponseResult(@NonNull flutterResult: Result, @NonNull result: RestResponse, code: String = ""){
+
+    var restResponse = FlutterRestResponse(result)
+
+    // if code is not 200 then throw an exception
+    if(!result.code.isSuccessful){
+      prepareError(flutterResult, Exception("RestAPI Exception Code " + result.code.toString()), "RestAPI error code received", code)
+      return
+    }
+    else {
+      Handler(Looper.getMainLooper()).post {
+        flutterResult.success(restResponse.toValueMap())
+      }
+      if (!code.isNullOrEmpty()) removeCode(code)
+    }
+  }
+
+  private fun prepareError(@NonNull flutterResult: Result, @NonNull error: Exception, @NonNull msg: String, code: String = "") {
+
+    var recoverySuggestion : String? = null
+    if(error is AmplifyException){
+      recoverySuggestion = "\nRecovery Suggestion: " + error.recoverySuggestion
+    }
+
+    Handler (Looper.getMainLooper()).post {
+      flutterResult.error(
+              "AmplifyApiRestException: " + error.cause.toString(),
+              msg,
+              error.message + (recoverySuggestion ?: "")
+      )
+    }
+
+    if(!code.isNullOrEmpty()) removeCode(code)
+  }
+
+  private fun checkArguments(@NonNull args: Map<String, Any>): Map<String, Any>{
+    if (args !is Map<String, Any>) {
+      throw java.lang.Exception("Flutter method call arguments must be a Map<String, dynamic>")
+    }
+
+    var inputMap : Map<String, Any> = args as Map<String, Any>
+
+    if (!inputMap.containsKey("restOptions") ||
+            inputMap["restOptions"] !is Map<*, *> ) {
+      throw java.lang.Exception("Flutter method call map does not contain valid RestOptions")
+    }
+
+    if (!inputMap.containsKey("code")){
+      throw java.lang.Exception("Flutter method call map does not contain valid code")
+    }
+
+    return inputMap
+  }
+
+  private fun removeCode(code: String){
+    if(operationsMap.containsKey(code)){
+      operationsMap.remove(code)
+    }
+  }
+
+  @VisibleForTesting
+  private fun onCancel(
+          @NonNull flutterResult: Result,
+          @NonNull code: String){
+
+    if(operationsMap.containsKey(code)){
+      operationsMap[code]!!.cancel()
+      operationsMap.remove(code)
+      flutterResult.success("Operation Canceled")
+    }
+    else{
+      flutterResult.error(
+              "Cancel - RestOperation referenced with code not found",
+              "The RestOperation may have already completed or expired and cannot be canceled anymore",
+              "Operation does not exist")
+    }
+  }
+
+  /*
+  GET
+  */
+  private fun get(
```

Comment:
Perhaps these could be split into a different file?

Reply:
^ Yea, that would be nice. The ideal architecture for _all_ of our platforms is to have the plugin be a thin veneer that just calls some ["service"](https://en.wikipedia.org/wiki/Service_(systems_architecture)) class. The service class is what actually _does_ "REST," the plugin is just a binding into the Amplify/Flutter system(s).

@raphkim did a nice job of this in the Android Predictions plugin, but unfortunately the pattern hasn't been used widely otherwise.

Reply:
Thanks guys we aim to do that in the merged branch and have been refactoring to reduce the amount of REST or GraphQL specific code within our main plugins file.